docs/prover/checking/sanity.md

```diff
@@ -2,9 +2,243 @@ Rule Sanity Checks
+There are 3 kinds of sanity checks:
+
+1. **Reachability** checks that even when ignoring all the user-provided
+   assertions, the end of the rule is reachable. This check ensures that that
+   the combination of `require` statements does not rule out all possible
+   counterexamples.
+
+   For example, the following rule would be flagged by the reachability check:
+   ```cvl
+   rule vacuous {
+     uint x;
+     require x > 2;
+     require x < 1;
+     assert f(x) == 2, "f must return 2";
+   }
+   ```
+   Since there are no models satisfying both `x > 2` and `x < 1`, this rule
+   will always pass, regardless of the behavior of the contract.  This is an
+   example of a *vacuous* rule - one that passes only because the preconditions
+   are contradictory.
+
+   ```{caution}
+   The reachability check will *pass* on vacuous rules and *fail* on correct
+   rules.  A passing reachability check indicates a potential error in the rule.
+   
+   The exception is when a {term}`parametric rule` is checked on the default
+   fallback function: The default fallback function should always revert, so
+   there are no examples that can reach the end of the rule.
+   ```
+   
+2. **Assert-Vacuity** checks that individual `assert` statements are not
+   tautologies.  A tautology is a statement that is true on all examples, even
+   if all the `require` and `if` conditions are removed.
+
+   For example, the following rule would be flagged by the assert-vacuity check:
+   
+   ```cvl
+   rule tautology {
+     uint x; uint y;
+     require x != y;
+     ...
+     assert x < 2 || x >= 2,
+      "x must be smaller than 2 or greater than or equal to 2";
+   }
+   ```
+  
+  
+   Since every `uint` satisfies the assertion, the assertion is tautological,
+   which is likely to be an error in the specification.
+   
+   **Checking vacuity for invariants**
+   
+   For invariant, vacuity is checked by converting it into a rule without any requires.
+
+    ```solidity
+     contract temp{
+     address root;
+     uint a;
+     constructor(address _root){
+         require(_root != address(0));
+         root = _root;
+     }
+     function changeRoot (address root_) public{
+         require(root_ != address(0));
+         root = root_;
+     }
+     }
+     ```
+   
+   
+    ```cvl
+    
+      // Not a tautology
+      invariant rootNonZero()
+         root() != 0
+
+      rule rootNonZeroRule(){
+         assert root() != 0;
+      }
+
+      // Tautology
+      invariant aGE0()
+         a() >= 0
+
+      rule aGE0Rule{
+         assert a() >= 0;
+      }
+
+    ```
+    
+    **Checking vacuity for rules**
+    
+    For rules, checking for tautology requires checking each assertion to see if 
+    it’s meaningful. In order to do this, we employ few different checks depending
+    on the syntax of the assertion expression.
+    
+      *Case 1: Implication*
+    
+      Given a rule with an `assert p => q` we perform two checks:
+    
+      1. `assert(!p)`
```

Comment:
let's give names to each check, so then in the ui (next release), we will give links to the doc from the error message
Implication Hypothesis